onnx/defs/sequence/defs.cc

```diff
@@ -0,0 +1,655 @@
+              target_dim.dim_param() == source_dim.dim_param()) {
+            return false;
+          }
+          return true;
+        }
+
+        return (target_dim.has_dim_value() || target_dim.has_dim_param());
+      }();
+      if (is_dims_conflict &&
+          (target_dim.has_dim_value() || target_dim.has_dim_param())) {
+        auto dim = target_shape->mutable_dim(i);
+        dim->clear_dim_value();
+        dim->clear_dim_param();
+      }
+    }
+  }
+}
+
+static const char* SequenceEmpty_ver11_doc = R"DOC(
+Construct an empty tensor sequence, with given data type.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SequenceEmpty,
+    11,
+    OpSchema()
+        .SetDoc(SequenceEmpty_ver11_doc)
+        .Attr(
+            "dtype",
+            "(Optional) The data type of the tensors in the output sequence. "
+            "The default type is 'float'.",
+            AttributeProto::INT,
+            OPTIONAL)
+        .Output(
+            0,
+            "output",
+            "Empty sequence.",
+            "S")
+        .TypeConstraint(
+            "S",
+            OpSchema::all_tensor_sequence_types(),
+            "Constrain output types to any tensor type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          const auto* attr_proto = ctx.getAttribute("dtype");
+          auto elem_type = TensorProto::FLOAT;
+          if (nullptr != attr_proto) {
+            if (!attr_proto->has_i()) {
+              fail_type_inference(
+                  "Attribute dtype should be of integer type and specify a type.");
+            }
+            auto attr_value = attr_proto->i();
+            elem_type = static_cast<TensorProto_DataType>(attr_value);
+          }
+          ctx.getOutputType(0)
+              ->mutable_sequence_type()
+              ->mutable_elem_type()
+              ->mutable_tensor_type()
+              ->set_elem_type(elem_type);
+        }));
+
+static const char* SequenceConstruct_ver11_doc = R"DOC(
+Construct a tensor sequence containing 'inputs' tensors.
+All tensors in 'inputs' must have the same data type.
+)DOC";
+
+ONNX_OPERATOR_SET_SCHEMA(
+    SequenceConstruct,
+    11,
+    OpSchema()
+        .SetDoc(SequenceConstruct_ver11_doc)
+        .Input(
+            0,
+            "inputs",
+            "List of tensors.",
+            "T",
+            OpSchema::Variadic)
+        .Output(
+            0,
+            "output_sequence",
+            "Sequence enclosing the input tensors.",
+            "S")
+        .TypeConstraint(
+            "T",
+            OpSchema::all_tensor_types(),
+            "Constrain input types to any tensor type.")
+        .TypeConstraint(
+            "S",
+            OpSchema::all_tensor_sequence_types(),
+            "Constrain output types to any tensor type.")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          size_t numInputs = ctx.getNumInputs();
+          if (numInputs < 1) {
+            return;
```

Comment:
If we have zero inputs, we will have a problem, since we cannot infer the output type (which is why "SequenceEmpty" needs a "dtype" attribute). The default interpretation for a variadic input is 1-or-more inputs. So, this case should not happen for a valid model. May be a comment that this cannot happen or fail_type_inference may be helpful.